Add: ConsumableItemBae 클래스 작성 및 BP 생성

wjddnjs6520 opened the pull request:
ConsumableItemBae
- OverlapCheckCollision에 BeginOverlap 시 캐릭터 쪽으로 아이템을 끌어당깁니다.
- PickupCheckCollision에 BeginOverlap 시 캐릭터 인벤토리에 해당 아이템이 없다면 추가 후 어빌리티 부여. 있다면 아이템의 숫자만 증가
- 캐릭터에 권한 부여 함수는 무기의 함수와 형태가 같습니다.
- DropItem은 적이 죽을 경우 아이템을 생성하기 위한 함수 입니다.
- MoveSpeed는 아이템이 캐릭터에게 다가오는 속도입니다.
- PlusItemCount는 아이템을 주울 시 인벤토리에서 증가할 개수입니다.

어빌리티 부여, 사용, 아이템 개수 감소 테스트를 진행하였고 PickupItem 함수에서 주석 처리된 부분과  ShooterCharacter에 임시로 아래 Map을 넣어 테스트 했습니다.
```cpp
UPROPERTY(EditAnywhere, BlueprintReadOnly, Category = "Test") 
TMap<TSubclassOf<AConsumableItembase>, int32> TestInventoryMap;
```
테스트를 위해서는 위 코드를 ShooterCharacter.h의 public 부분에 넣고 PickupItem의 주석 부분을 풀고 GA_UseItemBase 처음 부분 노드 링크를 이어주시면 되겠습니다.

또한 현재 구현한 PickupItem에 주석을 넣은 이유와 GA_UseItemBase 노드의 연결을 끊은 이유는 캐릭터의 인벤토리가 없기 때문이고 인벤토리가 만들어질 경우 구현 방법에 따라 수정할 예정입니다.

wjddnjs6520 committed:
Add: ConsumableItemBae 클래스 작성 및 BP 생성

ConsumableItemBae
- OverlapCheckCollision에 BeginOverlap 시 캐릭터 쪽으로 아이템을 끌어당깁니다.
- PickupCheckCollision에 BeginOverlap 시 캐릭터 인벤토리에 해당 아이템이 없다면 추가 후 어빌리티 부여. 있다면 아이템의 숫자만 증가
- 캐릭터에 권한 부여 함수는 무기의 함수와 형태가 같습니다.
- DropItem은 적이 죽을 경우 아이템을 생성하기 위한 함수 입니다.
- MoveSpeed는 아이템이 캐릭터에게 다가오는 속도입니다.
- PlusItemCount는 아이템을 주울 시 인벤토리에서 증가할 개수입니다.

어빌리티 부여, 사용, 아이템 개수 감소 테스트를 진행하였고 PickupItem 함수에서 주석 처리된 부분과  ShooterCharacter에 임시로 아래 Map을 넣어 테스트 했습니다.

UPROPERTY(EditAnywhere, BlueprintReadOnly, Category = "Test")
TMap<TSubclassOf<AConsumableItembase>, int32> TestInventoryMap;

테스트를 위해서는 위 코드를 ShooterCharacter.h의 public 부분에 넣고 PickupItem의 주석 부분을 풀고 GA_UseItemBase 부분 노드 링크를 이어주시면 되겠습니다.

hwarang98 commented on Source/Shooter/Private/ShooterGamePlayTag.cpp:
```diff
@@ -15,10 +15,11 @@ namespace ShooterGamePlayTags
 	UE_DEFINE_GAMEPLAY_TAG(InputTag_Weapon_Fire, "InputTag.Weapon.Fire");
 	UE_DEFINE_GAMEPLAY_TAG(InputTag_Weapon_Reload, "InputTag.Weapon.Reload");
-
+	UE_DEFINE_GAMEPLAY_TAG(InputTag_ConsumableItem_UseItemBase, "InputTag.ConsumableItem.UseItemBase");
```
해당 태그는 드랍 키 누를때 태그부여하는걸로 해석되는데 맞을까요?

wjddnjs6520 replied:
아이템 사용 인풋태그 입니다. 
ex) 힐 아이템 획득 -> 힐 아이템 사용 어빌리티, 인풋 태그 부여 -> **인풋태그 할당된 키 누르기** -> 힐 아이템 어빌리티 실행

hwarang98 replied:
아하 이해했습니다! 감사합니다!

hwarang98 commented on Source/Shooter/Private/Items/ConsumableItems/ConsumableItembase.cpp:
```diff
@@ -0,0 +1,140 @@
+
+        if (AbilityClass && OwnerASC->GetOwner()->HasAuthority())
+        {
+            FGameplayAbilitySpec Spec(AbilityClass, 1);
```
기본 생성자 FGameplayAbilitySpec(AbilityClass, 1)를 사용했는데, 이 경우 소유자 컴포넌트를 지정하지 않아 내부적으로 일부 동작에 문제가 있을 수 있어요

`FGameplayAbilitySpec Spec(AbilityClass, 1, INDEX_NONE, OwnerASC);` 이런 방식을 언리얼에서 권장하더라구여 이부분도 검토 부탁드립니다!

hwarang98 commented on Source/Shooter/Private/Items/ConsumableItems/ConsumableItembase.cpp:
```diff
@@ -0,0 +1,140 @@
+    UShooterAbilitySystemComponent* OwnerASC = NewOwner->GetShooterAbilitySystemComponent();
+    if (!OwnerASC) return;
+
+    for (const auto& Pair : ItemAbilities)
```
ItemAbilities 순회중 중복 부여 방지 로직도 추가되면 좋을것 같아요
`OwnerASC->FindAbilitySpecFromClass(AbilityClass)` 이걸로 조건걸고 이미 부여되있다면 스킵하는 로직도 추가되면 좋을것같은데 검토 부탁드립니다!

예시
```c++
       // 중복 부여 방지 예시
            if (OwnerASC->FindAbilitySpecFromClass(AbilityClass) != nullptr)
            {
                UE_LOG(LogTemp, Warning, TEXT("Ability %s already given"), *AbilityClass->GetName());
                continue;
            }
```

hwarang98 commented on Source/Shooter/Private/Items/ConsumableItems/ConsumableItembase.cpp:
```diff
@@ -0,0 +1,140 @@
+    UE_LOG(LogTemp, Warning, TEXT("MoveToCharacter"));
+    if (!Player) return;
+    SetActorTickEnabled(true); // 이동 시작 시 Tick 활성화
+    TargetLocation = Player->GetActorLocation();
```
제가 잘못이해한걸수도있는데 `TargetLocation = Player->GetActorLocation();` 이건 해당 함수 호출시 한번만 호출되는걸로 해석되는데 만약 제가 해석한게 맞다면 플레이어가 이동시 매 틱 마다 변해야하는데 한번만 세팅되는것 같은데 맞을까요?

wjddnjs6520 replied:
그 부분을 고려하지 않았던 것 같습니다. 처음에 오버랩되면 캐릭터를 targetactor로 설정하고 tick에서 계속 따라가게 변경하겠습니다.

hwarang98 replied:
넵 감사합니다!

wjddnjs6520 committed:
Fix: 플레이어 추적 위치 갱신 추가 및 코드 정리

플레이어가 한번 오버랩되면 PickupCheck 범위 안으로 들어올 때까지 지속 추적하도록 수정

권장되지 않는 코드 형식 변경
- weaponbase 코드도 포함해서 변경

어빌리티 중복 할당 방지 코드 추가
